doc/prov_loadex.md

```diff
@@ -0,0 +1,83 @@
+Providers run-time configuration
+================================
+
+Currently any provider run-time activation requires presence of the
+initialization parameters in the OpenSSL configuration file. Otherwise the
+provider will be activated with some "default" settings, that may or may not
+work for a particular application. For real-world systems it may require
+providing a specially designed OpenSSL config and passing it somehow (e.g. via
+environment) that has its obvious drawbacks.
+
+We need a possibility to initialize providers on per-application level
+according to per-application parameters. It's necessary for example for PKCS#11
+provider (where different applications may use different devices with different
+drivers) and will be useful for some other providers. In case of Red Hat it is
+also usable for FIPS provider.
+
+Real-world cases
+----------------
+
+Many applications use PKCS#11 API with a specific drivers. OpenSSL PKCS#11
+provider (https://github.com/latchset/pkcs11-provider) also provide a set of
+tweaks usable in particular situations. So there are at least several scenarios
+I have in mind:
+
+1. Configure a provider in the congig file, activate on demand
+2. Partially configure a provider in the config file, provide some extra
+   parameters later
+3. Load/activate a provider run-time and provide parameters later
+4. Load/activate a provider run-time with parameters
+5. Load a provider with a specific config snippet (usability)
+
+Proposed API
+------------
+
+All the names suggested below can be changed. I'm pretty bad in naming.
+
+#Configure a provider in the congig file, activate on demand
+
+We have an existing API `OSSL_PROVIDER_load`
+
+#Partially configure a provider in the config file, provide some extra
+parameters later
+
+We need an extra function
+
+```C
+int OSSL_PROVIDER_set_params(OSSL_PROVIDER *prov, OSSL_PARAM params[]);
+```
+
+and probably a syntax sugar
+
+```C
+int OSSL_PROVIDER_set_params_by_name(OSSL_LIB_CTX *libctx, const char *name,
+                                     OSSL_PARAM params[]);
+```
+
+#Load/activate a provider run-time and provide parameters later
+
+A pair of `OSSL_PROVIDER_load` and `OSSL_PROVIDER_set_params` should work
+
+#Load/activate a provider run-time with parameters
+
+```C
+OSSL_PROVIDER *OSSL_PROVIDER_load_ex(OSSL_LIB_CTX *libctx, const char *name,
+                                     OSSL_PARAM params[]);
+int OSSL_PROVIDER_load_with_params(OSSL_LIB_CTX *libctx, const char *name,
+                                   OSSL_PARAM params[]);
+```
+
+The rationale for the 2nd function is: we don't need a provider handle for most
+cases. We just want the provider to be loaded.
```

Comment:
This is trying to solve a different problem than "providing the provider parameters from an application". I am not sure it is a good idea to mix up these two problems together. But yeah, I agree that we perhaps unnecessarily require holding a provider reference when an application just wants a provider to be loaded into a library context for its whole lifetime.

Perhaps the function should be named OSSL_LIB_CTX_load_provider() instead?

Now the question is what should happen if the provider is already loaded, i.e. the provider's initialization already happened with different parameters (or no parameters at all)?

Reply:
I think we may want to deinit the provider and reinit from scratch ?
Potentially dangerous though.

Alternatively we return an error (already loaded), and the application can take it from there.

Reply:
@simo5 I think that there may be parameters that can be set in arbitrary moment of time. But provider may reject all the params change otherwise than on initialization.

Reply:
It is not how it works today, but you could explicitly propose behavior changes I guess?

Reply:
Yes.

Reply:
The file is removed from the PR, resolving the thread